pkg/collector/corechecks/snmp/internal/report/report_bandwidth_usage.go

```diff
@@ -97,18 +98,71 @@ func (ms *MetricSender) sendBandwidthUsageMetric(symbol profiledefinition.Symbol
 
-	sample := MetricSample{
-		value:      valuestore.ResultValue{SubmissionType: profiledefinition.ProfileMetricTypeCounter, Value: usageValue},
-		tags:       tags,
-		symbol:     profiledefinition.SymbolConfig{Name: usageName + ".rate"},
-		forcedType: profiledefinition.ProfileMetricTypeCounter,
-		options:    profiledefinition.MetricsConfigOption{},
+	interfaceID := ms.hostname + fullIndex + usageName
+	log.Debugf("interfaceID: %s", interfaceID)
+	err = ms.calculateRate(interfaceID, ifSpeed, usageValue, usageName, tags)
+	if err != nil {
+		return err
 	}
+	return nil
+}
 
-	ms.sendMetric(sample)
+/*
+calculateRate is responsible for checking the state for previously seen metric sample to generate the rate from.
+If the ifSpeed has changed for the interface, the rate will not be submitted (drop the previous sample)
+*/
+func (ms *MetricSender) calculateRate(interfaceID string, ifSpeed uint64, usageValue float64, usageName string, tags []string) error {
+	ms.interfaceRateMap.mu.Lock()
+	defer ms.interfaceRateMap.mu.Unlock()
+
+	interfaceRate, ok := ms.interfaceRateMap.rates[interfaceID]
+	log.Debugf("in the calculate rate, interfaceID: %s", interfaceID)
+
+	// current data point has the same interface speed as last data point
+	if ok && interfaceRate.ifSpeed == ifSpeed {
+		log.Debugf("in the calculate rate and in the if statement ok, interfaceID: %s", interfaceID)
+		// calculate the delta
+		currentTimestamp := TimeNow
+		delta := (usageValue - interfaceRate.previousSample) / (currentTimestamp - interfaceRate.previousTs)
+		// update the map previous as the current for next rate
+		interfaceRate.previousSample = usageValue
+		interfaceRate.previousTs = currentTimestamp
+		ms.interfaceRateMap.rates[interfaceID] = interfaceRate
+		// create the gauge metric sample
+		sample := MetricSample{
+			value:      valuestore.ResultValue{SubmissionType: profiledefinition.ProfileMetricTypeGauge, Value: delta},
+			tags:       tags,
+			symbol:     profiledefinition.SymbolConfig{Name: usageName + ".rate"},
+			forcedType: profiledefinition.ProfileMetricTypeGauge,
+			options:    profiledefinition.MetricsConfigOption{},
+		}
+		log.Debugf("send metric, interfaceID: %s, delta: %f", interfaceID, delta)
+		// send the metric
+		ms.sendMetric(sample)
+	} else {
+		// otherwise, no previous data point / different ifSpeed - make new entry for interface
+		ms.interfaceRateMap.rates[interfaceID] = InterfaceRate{
+			ifSpeed:        ifSpeed,
+			previousSample: usageValue,
+			previousTs:     TimeNow,
+		}
+		log.Debugf("new entry in interface map: interface ID: %s, ifSpeed: %d, previous sample: %f, ts: %f", interfaceID, ifSpeed, usageValue, TimeNow)
+		// do not send a sample to metrics, send error for ifSpeed change (previous entry conflicted)
+		if ok {
+			return fmt.Errorf("ifSpeed changed from %d to %d for device and interface %s, no rate emitted", interfaceRate.ifSpeed, ifSpeed, interfaceID)
```

Comment:
I like the idea of logging this, but I think it might make more sense as an `Info` level log. Interface speed changing is a normal thing to happen for some devices so I think an Info log is all we'd need here.

Might be nice to log the device IP/ID as well here so customers can get a better idea of exactly which device had this change happen.

Reply:
interface ID should encompass that! 

good point, even if i return an error, i realize this only bubbles up as a debug log later, i'll. update this as an info log!